[Validator] Add new `Schema` validation constraint

WedgeSama opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        |
| License       | MIT
|Doc PR      | https://github.com/symfony/symfony-docs/pull/20323


This constraint is like an extension of `Json` and `Yaml` constraints. It checks if given value is valid for the given format, then, if you declare `constraints` option, it will try to apply those constraints to the deserialized data.


Usage examples:

```php
// src/Entity/Report.php
namespace App\Entity;

use App\Yaml\MyConfiguration;
use Symfony\Component\Validator\Constraints as Assert;

class Report
{
    // Only validate format (like Json or Yaml constraints).
    #[Assert\Schema(
        format: 'yaml',
    )]
    private string $customConfiguration;

    // Same in short.
    #[Assert\Schema('yaml')]
    private string $customConfiguration;

    // Validate on parsed data, with simple constraints.
    #[Assert\Schema(
        format: 'yaml',
        constraints: new Assert\NotNull(),
    )]
    private string $customConfiguration;

    // Validate on parsed data, with composite constraints (example with `Collection`).
    #[Assert\Schema(
        format: 'yaml',
        constraints: new Assert\Collection([
            'foo' => new Assert\NotNull(),
        ]),
    )]
    private string $customConfiguration;
}
```

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/Yaml.php:
```diff
@@ -38,6 +40,9 @@ public function __construct(
         ?array $groups = null,
         mixed $payload = null,
+        public string $configMessage = 'This value do not match the required config.',
+        /** @var class-string<\Symfony\Component\Config\Definition\ConfigurationInterface>|null */
```
The php doc should be put on top of constructor, like the other ones

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/YamlValidator.php:
```diff
@@ -59,5 +64,28 @@ public function validate(mixed $value, Constraint $constraint): void
+
+        if (isset($data) && null !== $constraint->configClass) {
+            if (!interface_exists(ConfigurationInterface::class)) {
+                throw new LogicException('To use "configurationClass" option, you need to install "symfony/config" dependency.');
```
```suggestion
                throw new LogicException('The "configurationClass" option requires the "%s" interface. Try running "composer require symfony/config".', ConfigurationInterface::class);
```

To match other occurrences of messages of this type

xabbuh commented:
I doubt that this is a common enough use case that justifies to add this feature to the Symfony core.

WedgeSama commented:
IMO, now that there is a Yaml constraint, I think config validation will be common.

It allow to validate your Yaml tree directly (e.g. in form submit, or API call) without more action but to create your Configuration class.

That way, you ensure the Yaml can be validate soon in your process without to much code to do it.

IMO, it can be good to also add this to Json constraint too.

xabbuh commented:
I have never see an application where the config for a bundle is managed by a user of the application. Can you explain with more details how your actual use case looks like please?

WedgeSama commented:
The case I see is not about bundle's configuration (and you are right, bundle's config must not be managed by your app).

It is about configuration in general. In some app, you can find/manage config for other app or parts of the app itself, like [Home Assistant](https://www.home-assistant.io/) for example (just the principle).

In the case of Home Assistant, Yaml is used to configure and customize users' dashboard, directly from the web interface. And the problem for this example is the lack of validation on submit. With this PR, problem solve.

I know some app I worked on where this capability will have been good.

stof commented:
@WedgeSama I would suggest you to create a custom constraint in your project for that case (or in a third-party bundle if you want to reuse it). I don't think this use case is common enough to justify adding it in the core.

Btw, your proposal requires that the provided config class can be instantiated with no argument, which is not guaranteed at all by the interface.

WedgeSama commented:
After a chat with @Neirda24 it may be more relevant to use encapsulation of [Collection constraint](https://symfony.com/doc/current/reference/constraints/Collection.html) to do only validation. It will match more the logic of the Validator.

It will do something similar to what I wanted to do, but without `symfony/config` (it will also solve the config class instance you mention).

I'll update the PR.

WedgeSama committed:
Add new `Schema` validation constraint.
This constraint is like an extension of `Json` and `Yaml` constraints. It checks if given value is valid for the given format, then, if you declare `constraints` option, it will try to apply those constraints to the deserialized data.

alexandre-daubois commented:
If I want to validate a JSON against an actual schema, I would create a new constraint using, for example, https://packagist.org/packages/justinrainbow/json-schema. I'm not sure I would like to rewrite it with constraints by hand.

As far as I can tell, for the Yaml part, the PR only validates that the Yaml is valid, no schema is actually checked? Also, you're restoring an error handler in the validator but you didn't set one, what's the reason?

valtzu commented:
I recall having a similar need in the past, also related to config management via UI – so +1 for the idea from me.

---

Instead of putting serializer logic into validator component, how about adding a `#[Assert\Serialized(format: '...', constraints: [...])]` constraint to `symfony/serializer`?

And then in the framework config wire `@serializer` to the validator. This way we'd support all the formats that serializer supports. And of course some defaults would need to be configured in case you use the components without the framework.

wkania commented:
Perhaps it would be better if this was an optional `schema` definition within the JSON constraint?
Similar to how the `egulias/email-validator` is used for email validation, in this case we could use `justinrainbow/json-schema`.
